### PR TITLE
[mlir][sparse] Populate lvlToDim

### PR DESCRIPTION
**Pull request by yinying-lisa-li**
Updates:
1. Infer lvlToDim from dimToLvl
2. Add more tests for block sparsity
3. Finish TODOs related to lvlToDim, including adding lvlToDim to python binding

Verification of lvlToDim that user provides will be implemented in the next PR.

### CODE REVIEW

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 1]
```diff
@@ -811,7 +875,7 @@ RankedTensorType sparse_tensor::getCOOFromTypeWithOrdering(RankedTensorType rtt,
   // default value.
   unsigned posWidth = src.getPosWidth();
   unsigned crdWidth = src.getCrdWidth();
-  AffineMap invPerm; // TODO
+  auto invPerm = src.getLvlToDim();
```
Comment (PeimingLiu) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:878]: Revert the change here, do not use auto unless the type is obvious.

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 2]
```diff
@@ -749,6 +748,71 @@ mlir::sparse_tensor::getSparseTensorEncoding(Type type) {
...
+  lvlExprs.reserve(numLvls);
+  // lvlExprComponents stores information of the floordiv and mod operations
+  // applied to the same dimension, so as to build the lvlToDim map.
+  // Map key is the position of the dimension in dimToLvl.
+  // Map value is a SmallVector that contains lvl var for floordiv, multiplier,
+  // lvl var for mod in dimToLvl.
+  // For example, for il = i floordiv 2 and ii = i mod 2, the SmalleVector
+  // would be [il, 2, ii]. It could be used to build the AffineExpr
+  // i = il * 2 + ii in lvlToDim.
+  std::map<unsigned, SmallVector<AffineExpr, 3>> lvlExprComponents;
```
Comment (PeimingLiu) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:779]: This can be a vector right? Seems like you always index the map by dim.

Comment (PeimingLiu) [reply mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:779]: I would also (personally) prefer tuple here instead of `SmallVector`

File: mlir/lib/CAPI/Dialect/SparseTensor.cpp [hunk 3]
```diff
@@ -48,15 +48,17 @@ bool mlirAttributeIsASparseTensorEncodingAttr(MlirAttribute attr) {
...
+                                MlirAffineMap dimToLvl, MlirAffineMap lvlToDim,
+                                int posWidth, int crdWidth) {
   SmallVector<DimLevelType> cppLvlTypes;
   cppLvlTypes.reserve(lvlRank);
   for (intptr_t l = 0; l < lvlRank; ++l)
     cppLvlTypes.push_back(static_cast<DimLevelType>(lvlTypes[l]));
-  mlir::AffineMap lvlToDim; // TODO: provide in API
+  auto unwrappedLvlToDim = unwrap(lvlToDim);
+  if (!unwrappedLvlToDim)
+    unwrappedLvlToDim = inferLvlToDim(unwrap(dimToLvl), unwrap(ctx));
```
Comment (aartbik) [inline mlir/lib/CAPI/Dialect/SparseTensor.cpp:59]: I am not sure if we should infer inside the CAPI construction (which is really a binding to the outside world), or whether this should just mechanically accept whatever is given. If we do inference during construction, if feels this logic should be in SparseTensorEncodingAttr::get itself, not in the binding code

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 4]
```diff
@@ -749,6 +748,71 @@ mlir::sparse_tensor::getSparseTensorEncoding(Type type) {
...
+  }
+  return lvlToDim;
+}
+
+AffineMap mlir::sparse_tensor::inverseBlockSparsity(AffineMap dimToLvl,
+                                                    MLIRContext *context) {
+  SmallVector<AffineExpr> lvlExprs;
+  auto numLvls = dimToLvl.getNumResults();
+  lvlExprs.reserve(numLvls);
+  // lvlExprComponents stores information of the floordiv and mod operations
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:771]: I would move some of the comments into the code rather than all up from (especially the floordiv/mod explanations)

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp (hunk 2 above)
Comment (yinying-lisa-li) [reply mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:779]: I tried the vector way, but it's less convenient than map for building the lvlExprs in the for loop. If we use map, it could be empty and hence be skipped. But if we use vector, since we need to resize the vector in the beginning, it won't skip the loop. And it could lead to cases like below which are more complicated to deal with for resizing. The vector size should be 1, but it's not dimension size nor results size.
```
#NV_24 = #sparse_tensor.encoding<{
  map = ( i, j ) ->
  ( i            : dense,
    j floordiv 4 : dense,
    j mod 4      : block2_4
  )
}>
```

The similar goes to tuple vs SmallVector: for SmallVector, the assertion could be components.second.size() == 3, but if we switch to tuple, we would have to assert three things != Null. And I personally don't like c++'s cumbersome way of updating tuple: std::get<1>(components) = binOp.getRHS(). Since based on our previous conversation about this, you don't have a strong preference, I'll keep SmallVector. ;)

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensor.h [hunk 5]
```diff
@@ -159,6 +159,13 @@ inline bool hasAnySparseOperandOrResult(Operation *op) {
   return hasAnySparseOperand(op) || hasAnySparseResult(op);
 }
 
+//
+// Inference.
+//
+
+AffineMap inferLvlToDim(AffineMap dimToLvl, MLIRContext *context);
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensor.h:166]: Add doc to both methods, for example

/// Given the dimToLvl map, infers the lvlToDim map, or returns empty Affine when inference fails.

and

/// Returns the lvlToDim map for the given dimToLvl map specific to the block sparse cases. Asserts on failure (so only use when known to succeed).

Or something like that

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 6]
```diff
@@ -749,6 +749,74 @@ mlir::sparse_tensor::getSparseTensorEncoding(Type type) {
...
+AffineMap mlir::sparse_tensor::inferLvlToDim(AffineMap dimToLvl,
+                                             MLIRContext *context) {
+  auto map = static_cast<AffineMap>(dimToLvl);
+  AffineMap lvlToDim;
+  // Return an empty lvlToDim when inference is not successful.
+  if (!map || map.getNumSymbols() != 0) {
+    lvlToDim = AffineMap();
+  } else if (map.isPermutation()) {
+    lvlToDim = inversePermutation(map);
+  } else {
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:761]: Perhaps add a TODO to make this

else if (map.isBlock()) // pseudo-code but you get the idea
  lvlToDim = inverseBlockSparsity(map, context);
} else {
  // handle the error somehow
}

**Review by aartbik — APPROVED** [review]

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp (hunk 6 above)
Comment (yinying-lisa-li) [reply mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:761]: Yes, I have that in the verification PR.
